testing/base/source_gen_test.cpp

```diff
@@ -0,0 +1,173 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "testing/base/source_gen.h"
+
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+
+#include "common/set.h"
+#include "testing/base/gtest_main.h"
+#include "toolchain/driver/driver.h"
+
+namespace Carbon::Testing {
+namespace {
+
+using ::testing::AllOf;
+using ::testing::ContainerEq;
+using ::testing::Contains;
+using ::testing::Eq;
+using ::testing::Ge;
+using ::testing::Le;
+using ::testing::MatchesRegex;
+using ::testing::SizeIs;
+
+template <typename T>
+static auto SumSizes(const T& range) -> ssize_t {
+  return std::accumulate(
+      range.begin(), range.end(), static_cast<ssize_t>(0),
+      [](ssize_t lhs, const auto& rhs) -> ssize_t { return lhs + rhs.size(); });
+}
+
+TEST(SourceGenTest, UniqueIds) {
+  SourceGen gen;
+
+  auto unique = gen.GetShuffledUniqueIds(1000);
+  EXPECT_THAT(unique.size(), Eq(1000));
+  Set<llvm::StringRef> set;
+  for (llvm::StringRef id : unique) {
+    EXPECT_THAT(id, MatchesRegex("[A-Za-z][A-Za-z0-9_]*"));
+    EXPECT_TRUE(set.Insert(id).is_inserted()) << "Colliding id: " << id;
+  }
+
+  // Check that repeated calls are different in interesting ways, but have the
+  // exact same total bytes.
+  ssize_t unique_size_sum = SumSizes(unique);
+  for ([[maybe_unused]] int i : llvm::seq(1, 10)) {
```

Comment:
I think it'd be clearer about the number of loops if you don't pass a lower bound:

```suggestion
  for ([[maybe_unused]] int _ : llvm::seq(9)) {
```

Reply:
Dropped the lower bound, that was a mistake anyways.

But not using `_` intentionally as GoogleTest also uses an identifier with that name and it seems too clever to use it in loops instead of just marking the loop variable as unused....